File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/web/ExtendedRequestContextFilter.java
package com.github.arhor.simple.expense.tracker.web;

import com.github.arhor.simple.expense.tracker.util.JavaLangExt;
import lombok.experimental.ExtensionMethod;
import lombok.val;
import org.slf4j.MDC;
import org.springframework.boot.web.servlet.filter.OrderedRequestContextFilter;
import org.springframework.stereotype.Component;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Optional;
import java.util.UUID;

@Component
@ExtensionMethod(JavaLangExt.class)
@SuppressWarnings("NullableProblems")
public class ExtendedRequestContextFilter extends OrderedRequestContextFilter {

    private static final String REQUEST_ID = "X-REQUEST-ID";

    @Override
    protected void doFilterInternal(
        final HttpServletRequest request,
        final HttpServletResponse response,
        final FilterChain filterChain
    ) throws ServletException, IOException {
        super.doFilterInternal(request, response, wrap(filterChain));
    }

    private FilterChain wrap(final FilterChain filterChain) {
        return (req, res) -> {
            if ((req instanceof HttpServletRequest servletReq) && (res instanceof HttpServletResponse servletRes)) {
                val requestId = Optional.ofNullable(servletReq.getHeader(REQUEST_ID))
                    .map(UUID::fromString)
                    .orElseGet(UUID::randomUUID)
                    .also(CurrentRequestContext::new)
                    .toString();

                MDC.put("request-id", requestId);
                servletRes.addHeader(REQUEST_ID, requestId);
            }
            try {
                filterChain.doFilter(req, res);
            } finally {
                MDC.clear();
            }
        };
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/web/ExtendedRequestContextFilter.kt
package com.github.arhor.simple.expense.tracker.web

import org.slf4j.MDC
import org.springframework.boot.web.servlet.filter.OrderedRequestContextFilter
import org.springframework.stereotype.Component
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class ExtendedRequestContextFilter : OrderedRequestContextFilter() {

    override fun doFilterInternal(req: HttpServletRequest, res: HttpServletResponse, next: FilterChain) {
        super.doFilterInternal(req, res, next.withContextExtension())
    }

    private fun FilterChain.withContextExtension() = FilterChain { req, res ->
        if ((req is HttpServletRequest) && (res is HttpServletResponse)) {
            val requestId = Optional.ofNullable(req.getHeader(REQUEST_ID))
                .map(UUID::fromString)
                .orElseGet(UUID::randomUUID)
                .also { CurrentRequestContext(it) }
                .toString()

            MDC.put("request-id", requestId)
            res.addHeader(REQUEST_ID, requestId)
        }
        try {
            doFilter(req, res)
        } finally {
            MDC.clear()
        }
    }

    companion object {
        private const val REQUEST_ID = "X-REQUEST-ID"
    }
}
